cou/steps/analyze.py

```diff
@@ -0,0 +1,200 @@
+    return ""
+
+
+def check_os_versions(os_versions: Dict[str, defaultdict[Any, Set]]) -> Tuple:
+    """Check the consistency of OpenStack version on the deployment."""
+    versions = defaultdict(set)
+    upgrade_units = defaultdict(set)
+    upgrade_charms = defaultdict(set)
+
+    for app, os_release_units in os_versions.items():
+        os_version_units = set(os_release_units.keys())
+        for os_version_unit in os_version_units:
+            versions[os_version_unit].add(app)
+        if len(os_version_units) > 1:
+            logging.warning("Units from %s are not in the same openstack version", app)
+            os_sequence = sorted(
+                list(os_version_units),
+                key=lambda release: CompareOpenStack(release),  # pylint: disable=W0108
+            )
+            for os_release in os_sequence[:-1]:
+                next_release = determine_next_openstack_release(os_release)[1]
+                upgrade_units[next_release].update(os_release_units[os_release])
+                logging.warning(
+                    "upgrade units: %s from: %s to %s",
+                    os_release_units[os_release],
+                    os_release,
+                    next_release,
+                )
+
+    if len(versions) > 1:
+        logging.warning("Charms are not in the same openstack version")
+        os_sequence = sorted(
+            versions.keys(), key=lambda release: CompareOpenStack(release)  # pylint: disable=W0108
+        )
+        for os_release in os_sequence[:-1]:
+            next_release = determine_next_openstack_release(os_release)[1]
+            upgrade_charms[next_release].update(versions[os_release])
+            logging.warning(
+                "upgrade charm: %s from: %s to %s", versions[os_release], os_release, next_release
+            )
+
+    else:
+        actual_release = list(versions)[0]
+        next_release = determine_next_openstack_release(actual_release)[1]
+        logging.info(
+            "Charms are in the same openstack version and can be upgrade from: %s to: %s",
+            actual_release,
+            next_release,
+        )
+        upgrade_charms[next_release].update(os_versions.keys())
+
+    return upgrade_units, upgrade_charms
+
+
+def check_os_channels_and_migration(os_versions: Dict[str, defaultdict[Any, Set]]) -> Tuple:
+    """Check if applications has the right channel set and if needs charmhub migration."""
+    change_channel = defaultdict(set)
+    charmhub_migration = defaultdict(set)
+
+    for app in os_versions.keys():
+        app_channel, charm_origin = extract_app_channel_and_origin(app)
+        actual_release = list(os_versions[app].keys())
+        if len(actual_release) > 1:
+            logging.warning(
+                "Skip check of channels. App %s has units with different openstack version",
+                app,
+            )
+        else:
+            actual_release = actual_release[0]
```

Comment:
To avoid nesting too deeply.
```suggestion
        if len(actual_release) > 1:
            logging.warning(
                "Skip check of channels. App %s has units with different openstack version",
                app,
            )
            continue

        actual_release = actual_release[0]
```